File: landing/lib/phonology.ts
const VOWELS = new Set('aeiouáéíóúâêôãõàü')
const STRONG = new Set('aeoáéóâêô')
const DIGRAPHS = ['ch', 'lh', 'nh', 'qu', 'gu']
const CLUSTERS = [
  'bl', 'br', 'cl', 'cr', 'dl', 'dr', 'fl', 'fr',
  'gl', 'gr', 'pl', 'pr', 'tl', 'tr', 'vl', 'vr'
]

function isVowel(char: string): boolean {
  return VOWELS.has(char.toLowerCase())
}

function isHiatus(pair: string): boolean {
  if (pair.length < 2) return false
  return STRONG.has(pair[pair.length - 2]!) && STRONG.has(pair[pair.length - 1]!)
}

function onsetStart(run: string): number {
  if (run.length <= 1) return 0
  const tail = run.slice(-2)
  if (DIGRAPHS.includes(tail) || CLUSTERS.includes(tail)) return run.length - 2
  return run.length - 1
}

export function syllables(word: string): string[] {
  const text = word.trim()
  if (!text) return []

  const lowered = text.toLowerCase()
  const breaks: number[] = []
  let i = 0

  while (i < lowered.length) {
    if (!isVowel(lowered[i]!)) {
      i += 1
      continue
    }

    let j = i + 1
    while (j < lowered.length && isVowel(lowered[j]!) && !isHiatus(lowered.slice(i, j + 1))) j += 1

    let consonants = j
    while (consonants < lowered.length && !isVowel(lowered[consonants]!)) consonants += 1

    if (consonants >= lowered.length) break

    breaks.push(j + onsetStart(lowered.slice(j, consonants)))
    i = consonants
  }

  const pieces: string[] = []
  let previous = 0
  for (const point of breaks) {
    if (point > previous) {
      pieces.push(text.slice(previous, point))
      previous = point
    }
  }
  pieces.push(text.slice(previous))

  return pieces.filter(Boolean)
}

export function firstSyllable(word: string): string {
  const parts = syllables(word)
  if (parts.length === 0) return word.slice(0, 2)
  if (parts[0]!.length === 1 && parts.length > 2) return parts[0]! + parts[1]!
  return parts[0]!
}

File: ingest/steps/phonology.py
import unicodedata

VOWELS = set("aeiouáéíóúâêôãõàü")
DIGRAPHS = ("ch", "lh", "nh", "qu", "gu")
CLUSTERS = (
    "bl", "br", "cl", "cr", "dl", "dr", "fl", "fr",
    "gl", "gr", "pl", "pr", "tl", "tr", "vl", "vr"
)
SPLIT_PAIRS = ("rr", "ss", "sc", "sç", "xc", "xs")


def _is_vowel(char: str) -> bool:
    return char.lower() in VOWELS


def syllables(word: str) -> list[str]:
    text = word.strip()
    if not text:
        return []

    lowered = text.lower()
    breaks: list[int] = []
    i = 0

    while i < len(lowered):
        if not _is_vowel(lowered[i]):
            i += 1
            continue

        j = i + 1
        while j < len(lowered) and _is_vowel(lowered[j]) and _hiatus(lowered[i:j + 1]) is False:
            j += 1

        consonants = j
        while consonants < len(lowered) and not _is_vowel(lowered[consonants]):
            consonants += 1

        run = lowered[j:consonants]
        if consonants >= len(lowered):
            break

        breaks.append(j + _onset_start(run))
        i = consonants

    pieces: list[str] = []
    previous = 0
    for point in breaks:
        if point > previous:
            pieces.append(text[previous:point])
            previous = point
    pieces.append(text[previous:])

    return [piece for piece in pieces if piece]


def _hiatus(pair: str) -> bool:
    if len(pair) < 2:
        return False
    first, second = pair[-2], pair[-1]
    strong = set("aeoáéóâêô")
    return first in strong and second in strong


def _onset_start(run: str) -> int:
    if not run:
        return 0
    if len(run) == 1:
        return 0

    tail = run[-2:]
    if tail in DIGRAPHS or tail in CLUSTERS:
        return len(run) - 2
    if tail in SPLIT_PAIRS:
        return len(run) - 1
    return len(run) - 1


def first_syllable(word: str) -> str:
    parts = syllables(word)
    if not parts:
        return word[:2]
    if len(parts[0]) == 1 and len(parts) > 2:
        return parts[0] + parts[1]
    return parts[0]


def slug(text: str) -> str:
    normalised = unicodedata.normalize("NFKD", text)
    stripped = "".join(c for c in normalised if not unicodedata.combining(c))
    return "".join(c if c.isalnum() else "_" for c in stripped.lower()).strip("_")
